src/chain/stateTransition/block/attestations.ts

```diff
@@ -23,61 +21,84 @@ import {
+// assert data.crosslink.parent_root == hash_tree_root(parent_crosslink)
+// assert data.crosslink.data_root == ZERO_HASH  # [to be removed in phase 1]
+// validate_indexed_attestation(state, convert_to_indexed(state, attestation))
+
 
-export function processAttestation(state: BeaconState, attestation: Attestation): BeaconState {
+export default function processAttestation(state: BeaconState, attestation: Attestation): void {
   const currentEpoch = getCurrentEpoch(state);
   const previousEpoch = getPreviousEpoch(state);
   const data = attestation.data;
   const attestationSlot = getAttestationDataSlot(state, data);
+  let ffgData, parentCrosslink;
   assert(
     attestationSlot + MIN_ATTESTATION_INCLUSION_DELAY <= state.slot &&
     state.slot <= attestationSlot + SLOTS_PER_EPOCH
   );
 
-  // Check target epoch, source epoch, and source root
-  assert((
-    currentEpoch === data.targetEpoch &&
-    state.currentJustifiedEpoch === data.sourceEpoch &&
-    state.currentJustifiedRoot.equals(data.sourceRoot) &&
-    hashTreeRoot(state.currentCrosslinks[data.shard], Crosslink).equals(data.previousCrosslinkRoot)
-  ) || (
-    previousEpoch === data.targetEpoch &&
-    state.previousJustifiedEpoch === data.sourceEpoch &&
-    state.previousJustifiedRoot.equals(data.sourceRoot) &&
-    hashTreeRoot(state.previousCrosslinks[data.shard], Crosslink).equals(data.previousCrosslinkRoot)
-  ));
-
-  // Check crosslink data
-  assert(data.crosslinkDataRoot.equals(ZERO_HASH)); // TO BE REMOVED IN PHASE 1
-
-  // Check signature and bitfields
-  assert(verifyIndexedAttestation(state, convertToIndexed(state, attestation)));
+  // assert([previousEpoch, currentEpoch].includes(data.targetEpoch));
 
   // Cache pending attestation
   const pendingAttestation: PendingAttestation = {
-    data,
+    data: data,
     aggregationBitfield: attestation.aggregationBitfield,
     inclusionDelay: state.slot - attestationSlot,
     proposerIndex: getBeaconProposerIndex(state),
   };
 
   if (data.targetEpoch === currentEpoch) {
+    ffgData = [state.currentJustifiedEpoch, state.currentJustifiedRoot, currentEpoch];
+    parentCrosslink = state.currentCrosslinks[data.crosslink.shard];
+
     state.currentEpochAttestations.push(pendingAttestation);
   } else {
+    ffgData = [state.previousJustifiedEpoch, state.previousJustifiedRoot, previousEpoch];
+    parentCrosslink = state.previousCrosslinks[data.crosslink.shard];
     state.previousEpochAttestations.push(pendingAttestation);
   }
-  return state;
-}
 
-export default function processAttestations(state: BeaconState, block: BeaconBlock): void {
-  assert(block.body.attestations.length <= MAX_ATTESTATIONS);
-  for (const attestation of block.body.attestations) {
-    processAttestation(state, attestation);
-  }
+  // Check FFG data, crosslink data, and signature
+  assert(ffgData === [data.sourceEpoch, data.sourceRoot, data.targetEpoch]);
```

Comment:
If i console log ffgData with right side value it's always same. But the comparing always gives false. what's the right way of comparing then?

Reply:
it compares it by reference. Trying JSON.stringify it

Reply:
maybe better is to `serialize(ffgData, [number64])`

Reply:
mark as TODO to use `ssz.equals`. It currently has a bug that effects this case, but we can use it soonish.

Reply:
I already have a branch where I replaced everything with equals :)

Reply:
> ffgData

but ffgData is array of `[number64, buffer, number64]`. So, should it be `serialize(ffgData, [number64])` ?

Reply:
oh, true, hmm, thats annoying, I don't think we can serialize array of different types, @wemeetagain ?

Reply:
No, we can either compare each array index separately, or turn the arrays into objects.

Reply:
ffgData would be a nice object type though